### DISCUSSION TITLE
How to post a new toot to all followers

### DISCUSSION THREAD

**Discussion by markcellus** [mastodon/mastodon#18389] [Q&A]
I'm attempting to make a post to Mastodon using my own server via ActivityPub protocol. Per the "The message" section of the [How to implement a basic ActivityPub Server](https://blog.joinmastodon.org/2018/06/how-to-implement-a-basic-activitypub-server/#the-message) docs, it says `https://www.w3.org/ns/activitystreams#Public` needs to be added to the `to` field. Is that all that's required? Or does the Activity need to specify each of my followers (ids) in order for my followers to see it? Thanks for any information that is provided!

**Comment by markcellus** [comment]
Ah, answered my own question. Essentially, I'll need to make a POST request to all inboxes of followers using the activity, per [ActivityPub Delivery docs](https://www.w3.org/TR/activitypub/#delivery):

> An activity is delivered to its targets (which are [actors](https://www.w3.org/TR/activitypub/#actors)) by first looking up the targets' inboxes and then posting the activity to those inboxes.

@Gargron Would doing this without the `https://www.w3.org/ns/activitystreams#Public` still result in all followers seeing the new toot in their feeds? If so, what does Mastodon use the `https://www.w3.org/ns/activitystreams#Public` in the `to` field for? Thanks for any information!

**Reply by nightpool** [reply]
@markcellus the `to` and `cc` fields are used to determine the visibility of the post. having `https://www.w3.org/ns/activitystreams#Public` in the to or cc field marks your post as public, which means it can be displayed in hashtags, search, the federated timeline, other users can boost it, etc. if you don't include a `to` or `cc` property and just send your posts to someone's inbox directly, then only those people would be able to read it. does that make sense?

Additionally, Mastodon supports the idea of ["shared inbox" delivery (7.1.3)](https://www.w3.org/TR/activitypub/#shared-inbox-delivery), which means that you only need to POST the activity only once per server, to the sharedInbox for that server. In that case, you'd want to use the ID of the `followers` collection that you've set up on your actor in the `to` field, so that Mastodon knows that it's a post to your followers. (along with `as:Public` if it's a public post)

**Reply by markcellus** [reply]
Yes, this makes a lot of sense @nightpool! Thank you. I'll use the shared inbox delivery method then, which is great. So if I understand correctly, with all irrelevant keys omitted, the `Create` activity object would have the following `to` value to post publicly and to followers on mastodon.social?:

```js
// ie. POST to https://mastodon.social/inbox 
{
    to: [
        'https://www.w3.org/ns/activitystreams#Public', 
        <actor_id_of_follower1>, 
        <actor_id_of_follower2>, 
       // ...actor ids of all other followers here
    ]
}
```
